File: 3744-minimum-operations-to-make-array-elements-zero/3744-minimum-operations-to-make-array-elements-zero.py
class Solution:
    def numOps(self, s: int, e: int) -> int:
        ops = 0
        pow = 1
        base = 4
        last = s
        while base <= e:
            if base > last:
                ops += (base - last) * pow
                last = base
            base <<= 2
            pow += 1
        ops += (e - last + 1) * pow
        return ceil(ops / 2)
    def minOperations(self, queries: List[List[int]]) -> int:
        res = 0
        for s, e in queries:
            res += self.numOps(s, e)
        return res
        

File: 3744-minimum-operations-to-make-array-elements-zero/3744-minimum-operations-to-make-array-elements-zero.cpp
class Solution {
public:
    long long minOperations(vector<vector<int>>& queries) {
        long long res = 0;
        for(int i = 0; i < queries.size(); i++)
        {
            res += numOperations(queries[i][0], queries[i][1]);
        }
        return res;
    }
private:
    long long numOperations(int s, int e)
    {
        char pow = 1;
        int base = 4;
        int last = s;
        long long ops = 0;
        while(base <= e)
        {
            if(base > last)
            {
                ops += ((long long)base - last) * pow;
                last = base;
            }
            base <<= 2;
            pow++;
        }
        ops += ((long long)e - last + 1) * pow;
        return ceil((double)ops / 2);
    }
};